dimension.go

```diff
@@ -353,131 +353,53 @@ func (d *Dimension) Domain() (interface{}, error) {
-		ret = C.tiledb_dimension_get_domain(d.context.tiledbContext, d.tiledbDimension, &cdomain)
-		tmpslice := (*[1 << 46]C.uint16_t)(unsafe.Pointer(cdomain))[:2:2]
-		for i, s := range tmpslice {
-			tmpDomain[i] = uint16(s)
-		}
-		domain = tmpDomain
+		return domainInternal[uint16](d)
 	case TILEDB_UINT32:
-		cdomain := C.malloc(2 * C.sizeof_uint32_t)
-		defer C.free(cdomain)
-		tmpDomain := make([]uint32, 2)
-		ret = C.tiledb_dimension_get_domain(d.context.tiledbContext, d.tiledbDimension, &cdomain)
-		tmpslice := (*[1 << 46]C.uint32_t)(unsafe.Pointer(cdomain))[:2:2]
-		for i, s := range tmpslice {
-			tmpDomain[i] = uint32(s)
-		}
-		domain = tmpDomain
+		return domainInternal[uint32](d)
 	case TILEDB_UINT64:
-		cdomain := C.malloc(2 * C.sizeof_uint64_t)
-		defer C.free(cdomain)
-		tmpDomain := make([]uint64, 2)
-		ret = C.tiledb_dimension_get_domain(d.context.tiledbContext, d.tiledbDimension, &cdomain)
-		tmpslice := (*[1 << 46]C.uint64_t)(unsafe.Pointer(cdomain))[:2:2]
-		for i, s := range tmpslice {
-			tmpDomain[i] = uint64(s)
-		}
-		domain = tmpDomain
+		return domainInternal[uint64](d)
 	case TILEDB_FLOAT32:
-		cdomain := C.malloc(2 * C.sizeof_float)
-		defer C.free(cdomain)
-		tmpDomain := make([]float32, 2)
-		ret = C.tiledb_dimension_get_domain(d.context.tiledbContext, d.tiledbDimension, &cdomain)
-		tmpslice := (*[1 << 46]C.float)(unsafe.Pointer(cdomain))[:2:2]
-		for i, s := range tmpslice {
-			tmpDomain[i] = float32(s)
-		}
-		domain = tmpDomain
+		return domainInternal[float32](d)
 	case TILEDB_FLOAT64:
-		cdomain := C.malloc(2 * C.sizeof_double)
-		defer C.free(cdomain)
-		tmpDomain := make([]float64, 2)
-		ret = C.tiledb_dimension_get_domain(d.context.tiledbContext, d.tiledbDimension, &cdomain)
-		tmpslice := (*[1 << 46]C.double)(unsafe.Pointer(cdomain))[:2:2]
-		for i, s := range tmpslice {
-			tmpDomain[i] = float64(s)
-		}
-		domain = tmpDomain
+		return domainInternal[float64](d)
 	case TILEDB_BOOL:
-		cdomain := C.malloc(2 * C.sizeof_double)
-		defer C.free(cdomain)
-		tmpDomain := make([]bool, 2)
-		ret = C.tiledb_dimension_get_domain(d.context.tiledbContext, d.tiledbDimension, &cdomain)
-		tmpslice := (*[1 << 46]C.uint8_t)(unsafe.Pointer(cdomain))[:2:2]
-		for i, s := range tmpslice {
-			if s != 0 {
-				tmpDomain[i] = true
-			}
+		// Ensure that our booleans are in canonical true/false form in case they're
+		// a value other than 0 or 1.
+		asUints, err := domainInternal[uint8](d)
+		if err != nil {
+			return nil, err
 		}
-		domain = tmpDomain
+		return []bool{asUints[0] != 0, asUints[1] != 0}, nil
 	case TILEDB_STRING_ASCII:
-		domain = nil
-	default:
-		return nil, fmt.Errorf("Unrecognized domain type: %d", datatype)
+		return nil, nil
 	}
+	return nil, fmt.Errorf("Unrecognized domain type: %d", datatype)
+}
+
+func domainInternal[T any](d *Dimension) ([]T, error) {
+	// tiledb_dimension_get_domain writes *a pointer to the memory it owns*
+	// into cDomain, so we need to ensure that the domain stays alive for
```

Comment:
It may be unnecessary in a practical sense, but purely in the sense of wanting this function to be correct on its own, I want to ensure that it is not GC'd.

More importantly, there's nothing special about the receiver object in methods; it is allowed to get GC'd after it is last referenced, just like any other variable or function parameter. I put together [a toy example on the Go Playground](https://go.dev/play/p/8w--oEScEBX), where `val` is finalized before `val.Method()` is completed.